File: packages/palm-engine/memodb/memodb.go
// Package memodb encodes and decodes PalmOS Memo Pad's MemoDB.pdb
// records and AppInfo block.
//
// Each memo record is just a NUL-terminated text string. The memo's
// category is encoded in the low nibble of the record entry's
// attributes byte (0..15). The AppInfo block (286 bytes for MemoDB)
// carries the 16 category names + unique IDs.
//
// AppInfo layout for MemoDB (after the optional 2-byte pad that
// follows the record entry list):
//
//	0       2       renamedCategories bitmask (UInt16)
//	2       16*16   categoryLabels — NUL-padded ASCII names
//	258     16      categoryUniqIDs (UInt8 per category)
//	274     1       lastUniqueID (UInt8)
//	275     1       padding
//	276     2       reserved (UInt16)
//	278     2       sortByPriority (UInt16, 0 = manual, 1 = alphabetical)
//	         = 280 bytes total
package memodb

import (
	"encoding/binary"
	"errors"

	"github.com/palmvellum/palmvellum/packages/palm-engine/charset"
	"github.com/palmvellum/palmvellum/packages/palm-engine/pdb"
)

const (
	NumCategories   = 16
	CategoryNameLen = 16
	AppInfoLen      = 280
)

type Category struct {
	Name    string
	UniqID  uint8
	Renamed bool
}

type AppInfo struct {
	Categories     [NumCategories]Category
	LastUniqueID   uint8
	SortByPriority bool
}

// ParseAppInfo reads the categories AppInfo block.
func ParseAppInfo(b []byte) (*AppInfo, error) {
	if len(b) < 276 {
		return nil, errors.New("memodb: appinfo too short")
	}
	ai := &AppInfo{}
	renamed := binary.BigEndian.Uint16(b[0:2])
	for i := 0; i < NumCategories; i++ {
		off := 2 + i*CategoryNameLen
		end := off + CategoryNameLen
		// NUL-terminated string inside 16 bytes
		nameEnd := off
		for nameEnd < end && b[nameEnd] != 0 {
			nameEnd++
		}
		ai.Categories[i].Name = charset.FromPalm(b[off:nameEnd])
		ai.Categories[i].UniqID = b[2+NumCategories*CategoryNameLen+i]
		ai.Categories[i].Renamed = (renamed & (1 << i)) != 0
	}
	ai.LastUniqueID = b[2+NumCategories*CategoryNameLen+NumCategories]
	if len(b) >= 280 {
		// bytes 278..279 = sortByPriority
		ai.SortByPriority = binary.BigEndian.Uint16(b[278:280]) != 0
	}
	return ai, nil
}

// Encode emits the standard 280-byte MemoDB AppInfo block.
func (ai *AppInfo) Encode() []byte {
	out := make([]byte, AppInfoLen)
	var renamed uint16
	for i := range ai.Categories {
		if ai.Categories[i].Renamed {
			renamed |= 1 << i
		}
	}
	binary.BigEndian.PutUint16(out[0:2], renamed)
	for i, c := range ai.Categories {
		off := 2 + i*CategoryNameLen
		name := charset.ToPalm(c.Name)
		if len(name) > CategoryNameLen-1 {
			name = name[:CategoryNameLen-1]
		}
		copy(out[off:off+CategoryNameLen], name)
		out[2+NumCategories*CategoryNameLen+i] = c.UniqID
	}
	out[2+NumCategories*CategoryNameLen+NumCategories] = ai.LastUniqueID
	// 275 = padding
	// 276..277 = reserved
	if ai.SortByPriority {
		binary.BigEndian.PutUint16(out[278:280], 1)
	}
	return out
}

// FindCategoryByName returns the slot index (0..15) of the named
// category, or (0,false) if not present.
func (ai *AppInfo) FindCategoryByName(name string) (uint8, bool) {
	for i, c := range ai.Categories {
		if c.Name == name {
			return uint8(i), true
		}
	}
	return 0, false
}

// EnsureCategory makes sure a category with the given name exists.
// Returns its slot index. If a fresh slot is allocated, the AppInfo
// is mutated (caller should re-encode).
func (ai *AppInfo) EnsureCategory(name string) uint8 {
	if idx, ok := ai.FindCategoryByName(name); ok {
		return idx
	}
	// Find first empty slot (1..14; slot 0 is Unfiled, slot 15 is
	// system-reserved on real Palms).
	for i := 1; i < NumCategories-1; i++ {
		if ai.Categories[i].Name == "" {
			ai.LastUniqueID++
			ai.Categories[i].Name = name
			ai.Categories[i].UniqID = ai.LastUniqueID
			ai.Categories[i].Renamed = true
			return uint8(i)
		}
	}
	return 0 // out of slots — fall back to Unfiled
}

// CategoryName returns the human-readable name of a category index,
// or "Unfiled" for index 0 / unknown.
func (ai *AppInfo) CategoryName(idx uint8) string {
	if idx >= NumCategories {
		return "Unfiled"
	}
	name := ai.Categories[idx].Name
	if name == "" {
		return "Unfiled"
	}
	return name
}

// DefaultAppInfo returns an AppInfo seeded with the PalmOS factory
// categories — slot 0 "Unfiled", slot 1 "Personal", slot 2 "Business".
// All other slots are empty; EnsureCategory("AI") will land in slot 3.
func DefaultAppInfo() *AppInfo {
	ai := &AppInfo{}
	ai.Categories[0] = Category{Name: "Unfiled", UniqID: 0}
	ai.Categories[1] = Category{Name: "Personal", UniqID: 1}
	ai.Categories[2] = Category{Name: "Business", UniqID: 2}
	ai.LastUniqueID = 15
	return ai
}

// ─── Memo records ──────────────────────────────────────────────

type Memo struct {
	UniqueID uint32
	Category uint8
	Text     string
}

// DecodeMemos extracts all memo records from a parsed MemoDB.
func DecodeMemos(db *pdb.DB) []Memo {
	out := make([]Memo, 0, len(db.Records))
	for _, r := range db.Records {
		txt := r.Data
		// Trim trailing NULs (each memo is NUL-terminated)
		for len(txt) > 0 && txt[len(txt)-1] == 0 {
			txt = txt[:len(txt)-1]
		}
		out = append(out, Memo{
			UniqueID: r.UniqueID,
			Category: r.Attributes & 0x0F,
			Text:     charset.FromPalm(txt),
		})
	}
	return out
}

// EncodeMemos turns Memo structs into pdb.Record entries ready for
// pdb.DB.Write. Category is packed into the low nibble of attributes;
// each record's bytes are the memo text plus a single NUL terminator.
func EncodeMemos(memos []Memo) []pdb.Record {
	out := make([]pdb.Record, 0, len(memos))
	for _, m := range memos {
		enc := charset.ToPalm(m.Text)
		data := make([]byte, 0, len(enc)+1)
		data = append(data, enc...)
		data = append(data, 0)
		out = append(out, pdb.Record{
			UniqueID:   m.UniqueID,
			Attributes: m.Category & 0x0F,
			Data:       data,
		})
	}
	return out
}

// NewMemoDB constructs an empty MemoDB skeleton with the standard
// name/type/creator and the supplied AppInfo block already encoded.
func NewMemoDB(ai *AppInfo) *pdb.DB {
	return &pdb.DB{
		Name:    "MemoDB",
		Type:    pdb.FourCC("DATA"),
		Creator: pdb.FourCC("memo"),
		AppInfo: ai.Encode(),
	}
}

File: packages/android-native/app/src/main/java/dev/tatliving/palmvellum/organizers/data/card/MemoDb.kt
package dev.tatliving.palmvellum.organizers.data.card

/**
 * Encode/decode PalmOS Memo Pad's MemoDB records and AppInfo block — a Kotlin
 * port of `packages/palm-engine/memodb/memodb.go`.
 *
 * Each memo record is a NUL-terminated text string. The category is the low
 * nibble of the record-entry attributes byte (0..15). The AppInfo block carries
 * the 16 category names + unique IDs.
 *
 * AppInfo layout (280 bytes):
 *   0      2      renamedCategories bitmask (UInt16)
 *   2      16*16  categoryLabels (NUL-padded)
 *   258    16     categoryUniqIDs (UInt8 each)
 *   274    1      lastUniqueID
 *   275    1      padding
 *   276    2      reserved
 *   278    2      sortByPriority
 */

const val MEMO_NUM_CATEGORIES = 16
const val MEMO_CATEGORY_NAME_LEN = 16
const val MEMO_APPINFO_LEN = 280

class MemoCategory(
    var name: String = "",
    var uniqId: Int = 0,
    var renamed: Boolean = false,
)

class MemoAppInfo {
    val categories: Array<MemoCategory> = Array(MEMO_NUM_CATEGORIES) { MemoCategory() }
    var lastUniqueId: Int = 0
    var sortByPriority: Boolean = false

    fun findByName(name: String): Int? {
        for (i in categories.indices) if (categories[i].name == name) return i
        return null
    }

    /** Ensure a category exists; returns its slot index, mutating if a slot is allocated. */
    fun ensureCategory(name: String): Int {
        findByName(name)?.let { return it }
        // Slots 1..14 are user slots (0 = Unfiled, 15 system-reserved).
        for (i in 1 until MEMO_NUM_CATEGORIES - 1) {
            if (categories[i].name.isEmpty()) {
                lastUniqueId++
                categories[i].name = name
                categories[i].uniqId = lastUniqueId
                categories[i].renamed = true
                return i
            }
        }
        return 0 // out of slots -> Unfiled
    }

    fun categoryName(idx: Int): String {
        if (idx < 0 || idx >= MEMO_NUM_CATEGORIES) return "Unfiled"
        val n = categories[idx].name
        return if (n.isEmpty()) "Unfiled" else n
    }

    fun encode(): ByteArray {
        val out = ByteArray(MEMO_APPINFO_LEN)
        var renamed = 0
        for (i in categories.indices) if (categories[i].renamed) renamed = renamed or (1 shl i)
        putU16(out, 0, renamed)
        for (i in categories.indices) {
            val off = 2 + i * MEMO_CATEGORY_NAME_LEN
            var name = PalmCharset.toPalm(categories[i].name)
            if (name.size > MEMO_CATEGORY_NAME_LEN - 1) name = name.copyOf(MEMO_CATEGORY_NAME_LEN - 1)
            name.copyInto(out, off, 0, name.size)
            out[2 + MEMO_NUM_CATEGORIES * MEMO_CATEGORY_NAME_LEN + i] = categories[i].uniqId.toByte()
        }
        out[2 + MEMO_NUM_CATEGORIES * MEMO_CATEGORY_NAME_LEN + MEMO_NUM_CATEGORIES] = lastUniqueId.toByte()
        if (sortByPriority) putU16(out, 278, 1)
        return out
    }

    companion object {
        fun parse(b: ByteArray): MemoAppInfo? {
            if (b.size < 276) return null
            val ai = MemoAppInfo()
            val renamed = u16(b, 0)
            for (i in 0 until MEMO_NUM_CATEGORIES) {
                val off = 2 + i * MEMO_CATEGORY_NAME_LEN
                val end = off + MEMO_CATEGORY_NAME_LEN
                var nameEnd = off
                while (nameEnd < end && b[nameEnd].toInt() != 0) nameEnd++
                ai.categories[i].name = PalmCharset.fromPalm(b.copyOfRange(off, nameEnd))
                ai.categories[i].uniqId = b[2 + MEMO_NUM_CATEGORIES * MEMO_CATEGORY_NAME_LEN + i].toInt() and 0xFF
                ai.categories[i].renamed = (renamed and (1 shl i)) != 0
            }
            ai.lastUniqueId = b[2 + MEMO_NUM_CATEGORIES * MEMO_CATEGORY_NAME_LEN + MEMO_NUM_CATEGORIES].toInt() and 0xFF
            if (b.size >= 280) ai.sortByPriority = u16(b, 278) != 0
            return ai
        }

        /** PalmOS factory categories: Unfiled / Personal / Business. */
        fun default(): MemoAppInfo = MemoAppInfo().apply {
            categories[0] = MemoCategory("Unfiled", 0)
            categories[1] = MemoCategory("Personal", 1)
            categories[2] = MemoCategory("Business", 2)
            lastUniqueId = 15
        }

        private fun u16(b: ByteArray, o: Int) = ((b[o].toInt() and 0xFF) shl 8) or (b[o + 1].toInt() and 0xFF)
        private fun putU16(b: ByteArray, o: Int, v: Int) { b[o] = (v ushr 8).toByte(); b[o + 1] = v.toByte() }
    }
}

class Memo(
    var uniqueId: Int = 0,
    var category: Int = 0,
    var text: String = "",
)

object MemoDb {

    /** Decode all memo records from a parsed MemoDB. */
    fun decode(db: PalmDb): List<Memo> = db.records.map { r ->
        var txt = r.data
        var end = txt.size
        while (end > 0 && txt[end - 1].toInt() == 0) end--
        Memo(
            uniqueId = r.uniqueId,
            category = r.attributes and 0x0F,
            text = PalmCharset.fromPalm(txt.copyOfRange(0, end)),
        )
    }

    /** Turn memos into record entries: category in the low nibble, NUL-terminated bytes. */
    fun encode(memos: List<Memo>): MutableList<PalmRecord> = memos.map { m ->
        val enc = PalmCharset.toPalm(m.text)
        val data = ByteArray(enc.size + 1)
        enc.copyInto(data, 0, 0, enc.size) // trailing byte stays 0 (NUL terminator)
        PalmRecord(uniqueId = m.uniqueId, attributes = m.category and 0x0F, data = data)
    }.toMutableList()

    /** Empty MemoDB skeleton with the standard name/type/creator + encoded AppInfo. */
    fun newDb(ai: MemoAppInfo): PalmDb = PalmDb(
        name = "MemoDB",
        type = "DATA".toByteArray(Charsets.US_ASCII),
        creator = "memo".toByteArray(Charsets.US_ASCII),
        appInfo = ai.encode(),
    )
}
